### ISSUE TITLE
Design: last junior exiting during an outstanding loss windfalls the recovery to senior (junior-withdraw side of #145)

### ISSUE THREAD

**Issue by 0x-SquidSol** [dcccrypto/percolator-stake#161]
## Summary (design question, not a drain)

When the **last** junior LP fully exits while an insurance loss is **outstanding** (`total_flushed > total_returned`), their first-loss capital is zeroed out of the tranche, and a subsequent `ReturnInsurance` flows the recovered amount **entirely to senior** — even though senior never bore that loss (junior absorbed it). It is conservation-exact (no funds created or destroyed), so it is a cross-tranche **mis-allocation**, not a theft or drain. The mirror of the gated deposit side (#145/#150/#159): junior **deposits** during an outstanding loss are gated, but junior **withdrawals** are not, so a junior can crystallize their loss mid-claim and strand their own recovery.

We're filing this for a design decision rather than fixing it unilaterally, because the "right" behavior depends on the intended first-loss exit semantics.

## Affected code

- `src/processor.rs` — junior branch of `process_withdraw`: when the last junior LP burns (`new_junior_lp == 0`), raw `junior_balance` is set to 0 (the wipe at ~`:931-939`). There is no `total_flushed > total_returned` gate on the junior **withdraw** path (the gate added by #150 is on the **deposit** path only).
- `src/state.rs` — `effective_junior_balance()` returns 0 once `junior_balance == 0`, so `senior_balance() = total_pool_value() − 0` absorbs the entire `ReturnInsurance` credit.
- `src/processor.rs` — `process_return_insurance` credits `total_returned`, raising `total_pool_value()` (and thus `senior_balance()` when no junior remains).

## Mechanism + numbers

Tranches enabled, mode 0. Junior 100,000 (100,000 LP); senior 100,000 (100,000 LP). `total_deposited = 200,000`, `senior_balance = 100,000`.

1. Admin flushes **50,000** (junior-absorbed, 50,000 ≤ 100,000). `effective_junior_balance = 50,000`; `senior_balance = 100,000` (senior protected).
2. The (only) junior withdraws **all** 100,000 LP at the marked-down price → receives `calc_junior_collateral_for_withdraw(100_000, 50_000, 100_000) = 50,000` (realizes a −50,000 loss). `new_junior_lp == 0` → raw `junior_balance` set to **0**. `junior_total_lp = 0`.
3. After the exit: `total_pool_value = 100,000`, `effective_junior_balance = 0`, `senior_balance = 100,000` (senior still whole).
4. Admin returns **50,000** (`ReturnInsurance`): `total_returned = 50,000`, `net_loss = 0`, `total_pool_value = 150,000`. With no junior claimant, `senior_balance = 150,000` over 100,000 senior LP.
5. Senior withdraws 100,000 LP → **150,000** (deposited 100,000, **+50,000**). The +50,000 is exactly the insurance recovery of a loss **junior** absorbed and exited with.

Conservation holds: junior 50,000 + senior 150,000 = 200,000 deposited. The recovered 50,000 simply went to the wrong tranche (senior, which was protected) instead of back to the junior who bore it.

## Impact

A passive windfall to remaining senior LPs (and a forfeited recovery for an exiting junior) whenever the last junior exits during an outstanding, recoverable loss. Not unprivileged theft (it requires the junior to voluntarily exit mid-claim and the admin to return the insurance), and not a drain (conservation-exact). Severity is "medium / economic-fairness" — the concern is that **senior** ends up profiting from a loss it was structurally protected from.

Note the partial-exit variant is less clear-cut: if *some* junior remains, the recovery accrues to the **remaining juniors** (arguably fine — the exiting junior crystallized their loss). The sharp anomaly is the **last** junior exiting, which routes the recovery to senior.

## Proof of concept

State/math-level test using the real `calc_junior_collateral_for_withdraw` / `calc_senior_collateral_for_withdraw` + `StakePool::senior_balance` / `effective_junior_balance` / `total_pool_value`. Add as `tests/poc_junior_exit_windfall.rs`:

```rust
use bytemuck::Zeroable;
use percolator_stake::math::{calc_junior_collateral_for_withdraw, calc_senior_collateral_for_withdraw};
use percolator_stake::state::StakePool;

#[test]
fn last_junior_exit_during_loss_windfalls_recovery_to_senior() {
    let mut pool = StakePool::zeroed();
    pool.is_initialized = 1;
    pool.set_discriminator();
    pool.set_tranche_enabled(true);
    pool.total_deposited = 200_000;
    pool.total_lp_supply = 200_000;
    pool.set_junior_balance(100_000);
    pool.set_junior_total_lp(100_000);
    assert_eq!(pool.senior_balance().unwrap(), 100_000); // senior 100k / 100k LP

    // Junior-absorbed loss: senior protected.
    pool.total_flushed = 50_000;
    assert_eq!(pool.effective_junior_balance(), 50_000);
    assert_eq!(pool.senior_balance().unwrap(), 100_000);

    // Last junior exits all 100k LP at the marked-down price (gets 50k, realizes -50k).
    let j_out = calc_junior_collateral_for_withdraw(100_000, pool.effective_junior_balance(), 100_000).unwrap();
    assert_eq!(j_out, 50_000);
    pool.total_withdrawn += j_out;
    pool.total_lp_supply -= 100_000;
    pool.set_junior_total_lp(0);
    pool.set_junior_balance(0); // the wipe in process_withdraw
    assert_eq!(pool.senior_balance().unwrap(), 100_000);

    // Admin returns the insurance: no junior claimant -> windfalls to senior.
    pool.total_returned = 50_000;
    assert_eq!(pool.effective_junior_balance(), 0);
    assert_eq!(pool.total_pool_value().unwrap(), 150_000);
    assert_eq!(pool.senior_balance().unwrap(), 150_000);

    let s_out = calc_senior_collateral_for_withdraw(100_000, pool.senior_balance().unwrap(), 100_000).unwrap();
    assert_eq!(s_out, 150_000); // senior +50k = junior's forfeited recovery; conservation holds
}
```

**Reproduction** (verified against the v17 functions): junior absorbs and exits a 50,000 loss (−50,000); the 50,000 return then accrues entirely to senior (+50,000).

## Options for discussion

1. **Accept as-is** (document it): an exiting tranche LP crystallizes its loss and forfeits any future recovery; if no junior remains, the recovery accrues to senior. Simplest; matches "you exited, you're out." The wrinkle is that *senior* — the protected tranche — captures a junior-attributable recovery.
2. **Gate junior withdrawals during an outstanding loss** (mirror the deposit gate). Closes it symmetrically, but **traps junior funds** until the insurance is returned — a liveness cost, and the opposite direction of the HWM-freeze fix (#151/#152); likely undesirable.
3. **Preserve a junior claim on pending recovery after a full exit** — e.g. escrow the junior-attributable portion of an outstanding loss so a later return routes it back to (former) juniors rather than senior. Most "fair" but adds state/complexity.
4. **Cap senior's share of a return at its actual booked `senior_loss`** so a return can only restore senior up to what senior actually lost, with any excess (junior-attributable) handled separately. Targeted but needs careful accounting.

We'd defer to your call on which (if any) of these to pursue. Related: #145, #159 (deposit-side gates), #151/#152 (HWM-after-loss).

**Comment by dcccrypto** [comment]
**Decision (owner): do it perfectly fair.** → Option 4: cap senior's share of a `ReturnInsurance` at its actually-booked `senior_loss`, so a recovery can only restore senior up to what senior truly lost; any junior-attributable excess is NOT windfalled to senior (escrowed for former juniors or held in pool residual). NOT Option 2 (gating junior withdrawals would trap first-loss capital — opposite of the #151/#152 HWM-freeze fix). Needs careful accounting (track booked senior_loss; route excess). Implementation tracked as a follow-up.
